src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/InvokerUtil.cs

```diff
@@ -150,8 +168,74 @@ internal static InvokeDelegate GenerateInvokeDelegate(MethodInfo method, out int
 
                     return result;
                 };
+            }
+
+            private static InvokeDelegate GenerateInvokeDelegateInternalWithExpressions(int paramCount, bool returnsValue, int[] inputPos, int[] outputPos, MethodInfo method)
+            {
+                var targetParam = Expression.Parameter(typeof(object), "target");
+                var inputsParam = Expression.Parameter(typeof(object[]), "inputs");
+                var outputsParam = Expression.Parameter(typeof(object[]), "outputs");
+                var paramsLocal = Expression.Variable(typeof(object[]), "paramsLocal");
+                var result = Expression.Variable(typeof(object), "result");
+
+                var methodParam = Expression.Constant(method);
+
+                var assignParamsLocal = Expression.Assign(paramsLocal, Expression.Condition(
+                    Expression.GreaterThan(Expression.Constant(paramCount), Expression.Constant(0)),
+                    Expression.NewArrayBounds(typeof(object), Expression.Constant(paramCount)),
+                    Expression.Constant(null, typeof(object[]))));
+
+                var assignParamsLocalValues = inputPos.Length == 0
+                    ? Expression.Block(Expression.Empty())
+                    : Expression.Block(
+                        Enumerable.Range(0, inputPos.Length).Select(i =>
+                            Expression.Assign(
+                                Expression.ArrayAccess(paramsLocal, Expression.Constant(inputPos[i])),
+                                Expression.ArrayIndex(inputsParam, Expression.Constant(i)))));
+
+                var invokeMethod = Expression.Block(
+                    Expression.Assign(result, Expression.Condition(
```

Comment:
You can move the returnsValue check outside of the Expression creation and conditionally generate a different Expression based on it. No need to include in the execution of the Expression itself.

Reply:
done